Implement kNN for tf.idf similarity and use it for prediction

blanche opened the pull request:
also
- separate vectorization from similarity/knn computation
- use stopwords & lowercase in vectorizer
- use the same random seed for every model (still random, but better for comparison)
- output model name when printing error

blanche committed:
implement knn to find top K similar documents based on tf.idf vectors

blanche committed:
fix line-lengths >100

blanche committed:
fix bug where relevant_documents only contains the last relevant element, not a list

blanche committed:
implement weighted average prediction of knn based on tfidf.similarity

blanche committed:
rename cosine_knn function, cleanup function documentation

charlienewey committed:
Merge branch 'master' into standalone-similarity-impl

utkuozbulak commented:
Can you please not merge this just for a little more while, I we, I will check it as soon as possible.

blanche commented:
sure!
no rush for this one.. my next task is cleaning up the companies anyway

btw, the knn thing its pretty fast, even on 100k data set (i.e. < 5 min?), but i get memory error on my machine when running it on the full 120k, maybe have to play with the parallelization..

blanche committed:
Merge remote-tracking branch 'origin/master' into standalone-similarity-impl

blanche committed:
Merge remote-tracking branch 'origin/master' into standalone-similarity-impl

# Conflicts:
#	src/models/BaseModel.py
#	src/models/StandaloneSimilarity.py

blanche committed:
incorperate changes of OldMain / BaseModel merge
- add load all data to BaseModel __init__ because we need the raw description feature for
- remove KNN prallelization because of memory problems

blanche committed:
incorperate changes of OldMain / BaseModel merge
- add load all data to BaseModel __init__ because we need the raw description feature for
- remove KNN prallelization because of memory problems

blanche committed:
implement more memory efficient versions of Standalone Similarity

blanche commented:
done
error = 7.4k

utkuozbulak commented:
on full dataset ?

arahayrabedian commented on src/models/BaseModel.py:
```diff
@@ -38,6 +39,11 @@ def __init__(self, train_size=0.75, test_size=None):
             print('Pre-processed data exists, reading from the file')
             self.cleaned_encoded_data = self.load_cleaned_data()
             print('Data read complete')
+            if force_load_all:
```
could you summarize what's happening here? i'm a little lost as to why only `self.description_feature` is re-read.

blanche replied:
the problem is, that the `cleaned_encoded_data` does not contain the raw description data we need for the similarity model here. 

all other models are fine with the cleaned_encoded_data, so using the force_load_all you can force it to load (at least) the raw description data

arahayrabedian commented on src/models/StandaloneSimilarity.py:
```diff
@@ -1,27 +1,169 @@
 
+DATA_QUERIES_VECTOR_NPZ = '../preprocessing_data/queries_vector.npz'
+DATA_CORPUS_VECTOR_NPZ = '../preprocessing_data/corpus_vector.npz'
+DATA_SIMILIARITY_MATRIX = '../preprocessing_data/similiarity_matrix'
```
`.npz`? unused?

blanche replied:
npz is a numpy format
see https://docs.scipy.org/doc/numpy/reference/generated/numpy.savez.html

its very much used in models.StandaloneSimilarity.StandaloneSimilarity#store_precomputed_data

arahayrabedian commented on src/models/StandaloneSimilarity.py:
```diff
@@ -1,27 +1,169 @@
+FORCE_LOAD = True
+
 
 class StandaloneSimilarity(BaseModel):
```
i'mma just trust you on this chunk of code.

arahayrabedian commented on src/models/StandaloneSimilarity.py:
```diff
@@ -1,27 +1,169 @@
+        # prediction = self.custom_knn_calculation(prediction)
+
+        train_avg_salary = sum(self.y_train) / len(self.y_train)
+        cleaned_predictions = [x if str(x) != 'nan' else train_avg_salary for x in prediction]
```
there's got to be a better way to check for NaN

arahayrabedian replied:
also - average if nan? safe assumption? when does this happen?

blanche replied:
tbh i don't understand why other models don't have that problem, but it happens if you predict on the basis of other Nans (similar to having a 0 in a chain of multiplications)

better solution might be to clean input data better - should i give that a try?

if we cant make a proper prediction, we just fall back to the average salary

that being said i have only seen a single case of this in the entire data set - but the error calculation will fail if the predictions contain a nan value

arahayrabedian reviewed (commented):
not managing to internalize everything going on here, but looks generally not so bad

alexdy2007 committed:
Merge branch 'master' into standalone-similarity-impl
